### DISCUSSION TITLE
Call a function every time context is updated?

### DISCUSSION THREAD

**Discussion by n8jadams** [statelyai/xstate#2133] [Q&A] [answered]
It's pretty straightforward when interpreting a machine to have a callback run every time the state machine transitions:

```js
const promiseService = interpret(promiseMachine).onTransition(state =>
  console.log(state.value)
);
```

But is there a way to have a callback run whenever the context of the state machine is updated?

Something like

```js
const promiseService = interpret(promiseMachine).onContextChange(state =>
  publish(state.context)
);
```

**Comment by ChrisShank** [comment, marked as answer]
You were pretty close to guessing it! 

https://xstate.js.org/api/classes/interpreter.html#onchange

```ts
const promiseService = interpret(promiseMachine)

promiseService.onChange((context, prevContext) => {
  publish(context)
});
```

**Reply by n8jadams** [reply]
Thanks. Can't believe I missed that.

**Reply by davidkpiano** [reply]
Note: this (`service.onChange()` method) will be deprecated and removed in v5, but there are workarounds for it:

```js
// in v5
someService.subscribe(state => {
  if (state.context !== state.history?.context) {
    // context changed!
  }
});
```
